src/lib/WalletInfoCollector.ts

```diff
@@ -0,0 +1,214 @@
+        let initialAccountsPromise;
+        if (initialAccounts) {
+            WalletInfoCollector._addAccounts(walletInfo, initialAccounts, undefined, onIntermediateResult);
+            // after fetching balances, update again
+            initialAccountsPromise = WalletInfoCollector._getBalances(initialAccounts).then((balances) =>
+                WalletInfoCollector._addAccounts(walletInfo, initialAccounts, balances, onIntermediateResult));
+        } else {
+            initialAccountsPromise = Promise.resolve();
+        }
+
+        if (walletType === WalletType.LEGACY) {
+            // legacy wallets have no derived accounts
+            await initialAccountsPromise;
+            return walletInfo;
+        }
+
+        let didFindAccounts;
+        do {
+            const derivedAccounts = await derivedAccountsPromise;
+
+            // already start deriving next accounts
+            // By always advancing in groups of MAX_ALLOWED_GAP addresses per round, it often happens that more
+            // addresses are derived and checked for activity than the BIP44 address gap limit
+            // (https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#address-gap-limit) stipulates, because
+            // whenever an active address in a group of addresses is found, the next full group is also derived. Thus
+            // the actual gap limit of this implementation is up to (2 x MAX_ALLOWED_GAP) - 1.
+            // We argue that this is good UX for users, as potentially more of their active addresses are found, even
+            // if they haven't strictly followed to the standard - at only a relatively small cost to the network.
+            // For example, if the user adds the accounts derived with indices 0, 19, 39 to his wallet but then only
+            // ends up using accounts 0 and 39, the account at index 19 will not be found anymore on reimport. With the
+            // current implementation however, at least the account 39 would be found, while an implementation strictly
+            // following the specification would stop the search at index 20.
+            startIndex += WalletInfoCollector.MAX_ALLOWED_GAP;
+            derivedAccountsPromise = WalletInfoCollector._deriveAccounts(startIndex,
+                WalletInfoCollector.MAX_ALLOWED_GAP, walletType, walletId);
+
+            // find accounts with a balance > 0
+            didFindAccounts = false;
+            // TODO should use transaction receipts
+            const balances = await WalletInfoCollector._getBalances(derivedAccounts!);
+            const usedAccounts: BasicAccountInfo[] = [];
+            for (const account of derivedAccounts!) {
+                const balance = balances.get(account.address);
+                if (balance === undefined || balance === 0) continue;
+                usedAccounts.push(account);
+                didFindAccounts = true;
+            }
+            if (didFindAccounts) {
```

Comment:
Instead of `didFindAccounts`, you could use `usedAccounts.length > 0` and would get two flys with one stone ;)